opm/grid/cpgrid/Geometry.hpp

```diff
@@ -419,6 +419,59 @@ namespace Dune
                 return false;
             }
 
+            std::vector<Geometry<3, cdim>> refine(const std::array<int, 3>& cells,
+                                                  std::vector<EntityVariable<Geometry<0, cdim>, 3>>& corner_storage,
+                                                  std::vector<std::array<int, 8>>& indices_storage)
+            {
+                std::vector<Geometry<3, cdim>> result;
+                Geometry<3, cdim>::ctype volume = this->volume() / (cells[0] * cells[1] * cells[2]);
```

Comment:
You are correct, so far I was assuming rectangular shapes. However, if we assume general hexahedrons with 8 corners, it becomes complicated. I am not sure how to calculate the volume, and googling it did not give me an easy solution.

How is this handled so far in opm? Is there code for calculating the volume of a general hexahedron with 8 corners? Or is  there a sub-set of shapes that is supported?

Reply:
Personally I would split the hexahedron intotetrahedrons and calculate their volume. There is a known formula for that .

That seems to be what OPM is doing. in [opm/grid/cpgrid/processEclipseFormat.cpp](https://github.com/OPM/opm-grid/blob/bfeaf415de0d4e6246b48293baeff4da746b7364/opm/grid/cpgrid/processEclipseFormat.cpp#L1153-L1173). The tetrahedrons are spanned by the center of the cells and the points of a face. (You might want to double check my fast assumption, though).

Please be aware that the volumes might have changed later (deactivating small cells and adding their volume to a neighbor). Hence we might want to somehow make sure that the sum of the volumes is the total volume of the father cell by adding something to e.g. the largest cell.

Reply:
> Personally I would split the hexahedron into tetrahedrons and calculate their volume. There is a known formula for that.
> 
> That seems to be what OPM is doing. in [opm/grid/cpgrid/processEclipseFormat.cpp](https://github.com/OPM/opm-grid/blob/bfeaf415de0d4e6246b48293baeff4da746b7364/opm/grid/cpgrid/processEclipseFormat.cpp#L1153-L1173). The tetrahedrons are spanned by the center of the cells and the points of a face. (You might want to double check my fast assumption, though).

I looked at that code and found it way too complicated to understand easily. However, I did find the function for calculating the volume of a tetrahedron, so that makes it easier.

I used that to calculate the volume by splitting in 24 tetrahedrons, which appears to be a common approach. This seems to work fine. I still need to expand the tests a bit and then I will update the PR tomorrow (Thursday).

> Please be aware that the volumes might have changed later (deactivating small cells and adding their volume to a neighbor). Hence we might want to somehow make sure that the sum of the volumes is the total volume of the father cell by adding something to e.g. the largest cell.

Would it make sense to distribute the extra over all cells? By scaling each sub-volume with the ratio of the calculated total volume over the parent volume?

Reply:
@alfbr We are not changing the existing code. What we are talking about is calculating the volume of refined cells. If it turns out that something else is needed for restart, etc., we will change that. But it seem a good idea to do this iteratively and start with something simple. Hard to foresee what challenges we face further down the road.

Reply:
>  (assuming MINPV removes refined cells, which I have not double checked).

Somebody should probably check this, then. It certainly will make life more complicated.